src/main/java/org/kitteh/irc/client/library/implementation/STSHandler.java

```diff
@@ -0,0 +1,221 @@
+     * @param event the event instance
+     */
+    @Handler
+    public void onDisconnect(ClientConnectionClosedEvent event) {
+        // The spec says we have to update the expiry of the policy if it still exists
+        // at disconnection time...
+        // Do this by removing and re-adding.
+        String hostname = this.client.getConfig().getNotNull(Config.SERVER_ADDRESS).getHostName();
+        final STSStorageManager storageManager = this.machine.getStorageManager();
+        if (storageManager.hasEntry(hostname)) {
+            final STSPolicy policy = storageManager.getEntry(hostname).get();
+            long duration = Long.parseLong(policy.getOptions().get(STSPolicy.POLICY_OPTION_KEY_DURATION));
+            storageManager.removeEntry(hostname);
+            storageManager.addEntry(hostname, duration, policy);
+        }
+    }
+
+    private void handleSTSCapability(CapabilityState sts, List<ServerMessage> originalMessages) {
+        this.isSecure = client.getConfig().getNotNull(Config.SSL);
+        InetSocketAddress address = client.getConfig().getNotNull(Config.SERVER_ADDRESS);
+        final String msg = originalMessages.stream().map(ServerMessage::getMessage)
+            .collect(Collectors.joining()).replace('\n', ' ');
+        if (!sts.getValue().isPresent()) {
+            throw new KittehServerMessageException(msg, "No value provided for sts capability.");
+        }
+
+        final String capabilityValue = sts.getValue().get();
+        final STSPolicy policy = STSUtil.getSTSPolicyFromString(",", capabilityValue);
+        if (policy.getFlags().contains(STSPolicy.POLICY_OPTION_KEY_PORT) || policy.getFlags().contains(STSPolicy.POLICY_OPTION_KEY_DURATION)) {
+            throw new KittehServerMessageException(msg, "Improper use of flag in required option context!");
+        }
+
+        if (!policy.getOptions().containsKey(STSPolicy.POLICY_OPTION_KEY_PORT)) {
+            policy.getOptions().put(STSPolicy.POLICY_OPTION_KEY_PORT, Integer.toString(address.getPort())); // spec says port is optional
+        }
+
+        for (String key : policy.getOptions().keySet()) {
+            // Unknown keys are ignored by the switches below
+            this.machine.setSTSPolicy(policy);
+            if (this.isSecure) {
+                this.handleSecureKey(key, policy);
+            } else {
+                this.handleInsecureKey(key, policy);
+            }
+        }
+    }
+
+    private void handleInsecureKey(String key, STSPolicy policy) {
+        final String value = policy.getOptions().get(key);
+
+        switch (key) {
+            case STSPolicy.POLICY_OPTION_KEY_DURATION:
+                // Do NOT persist, because this policy could've been inserted by an active MitM
+                break;
+            case STSPolicy.POLICY_OPTION_KEY_PORT:
+                try {
+                    Integer.parseInt(value); // can't easily use a short because signed..
```

Comment:
It is, it's read in in the machine which has the input of the STS policy. It's not correct for the machine-specific logic to go in this event handler, but we should ensure we don't push the machine into a state for which we have an invalid policy/input - hence the validation here.